From 68e99204f671a814a3caca6bb4ac4cf125d98063 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=E5=BF=98=E6=80=80?= <guanwentao@uniontech.com>
Date: Tue, 24 Sep 2024 07:10:32 +0000
Subject: [PATCH 106/136] DEEPIN: HID: multitouch: Add quirk for HONOR
 MagicBook Art 14 touchpad
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

[Problem]
It sometimes after reboot change output from Touchpad to Mouse,
evtest show it output from "TOPS0102:00 35CC:0104 Touchpad"
to "TOPS0102:00 35CC:0104 Mouse",and it works as A mouse.

[Solution]
Like GLO-GXXX,As a workaround, it is possible to call
MT_QUIRK_FORCE_GET_FEATURE to force set feature in mt_set_input_mode for such special touchpad device.

Link:https://gitlab.freedesktop.org/libinput/libinput/-/issues/1040
Link:https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=9ffccb691adb854e7b7f3ee57fbbda12ff70533f
Link:https://www.cnblogs.com/szdytom/p/fix-honormagic-touchpad.html

Signed-off-by: 忘怀 <guanwentao@uniontech.com>
Link: https://github.com/deepin-community/kernel/commit/d61269a9d10654354c5a91eafdc230f8cca0329d
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 drivers/hid/hid-multitouch.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/hid/hid-multitouch.c b/drivers/hid/hid-multitouch.c
index 847462650549..6f555a83c6b6 100644
--- a/drivers/hid/hid-multitouch.c
+++ b/drivers/hid/hid-multitouch.c
@@ -2086,6 +2086,11 @@ static const struct hid_device_id mt_devices[] = {
 		HID_DEVICE(BUS_I2C, HID_GROUP_MULTITOUCH_WIN_8,
 			0x347d, 0x7853) },
 
+	/* HONOR MagicBook Art 14 touchpad */
+	{ .driver_data = MT_CLS_VTL,
+		HID_DEVICE(BUS_I2C, HID_GROUP_MULTITOUCH_WIN_8,
+			0x35cc, 0x0104) },
+
 	/* Ilitek dual touch panel */
 	{  .driver_data = MT_CLS_NSMU,
 		MT_USB_DEVICE(USB_VENDOR_ID_ILITEK,
-- 
2.47.0

